File: 49_GroupAnagrams.js


var groupAnagrams = function(strs) 
{
    const map = new Map();

    for(let i = 0; i < strs.length; i++)
    {
        let key = strs[i].split('').sort().join('');

        if(map.has(key)) {
            map.get(key).push(strs[i]);
        } else {
            map.set(key,[strs[i]]);
        }
    }
    console.log(map);
    const result = Array.from(map.values());
    return result;
};

var groupAnagrams1 = function(strs) 
{
    let result = {};

    for(let i = 0; i < strs.length; i++) 
    {
        let key = strs[i].split('').sort().join('');
        
        if(!result[key]) 
        {
            result[key] = [];
        }
        result[key].push(strs[i]);
    }
    return Object.values(result);
};

let main = function()
{
    let strs = ["eat","tea","tan","ate","nat","bat"];

    let result = groupAnagrams(strs);

    console.log(result);
}

main();

File: 49_GroupAnagrams.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>
#include<algorithm>
using namespace std;

class Solution 
{
    public:
        vector<vector<string>> groupAnagrams(vector<string>& strs) 
        {
            unordered_map<string, vector<string>> groups;            
            vector<vector<string>> result;

            for(const auto& str : strs) {
                vector<int> char_freq(26,0);

                for(const char& ch : str)             {
                    char_freq[ch - 'a']++;
                }
                string key = "";
                for(auto& freq : char_freq) {
                    key += to_string(freq) + "-";
                }
                groups[key].push_back(str);
                key.clear();
            }
            for(const auto& [key,values] : groups) {
                result.push_back(values);
            }
            return result;
        }
};

int main()
{
    vector<string>  strs = {"eat","tea","tan","ate","nat","bat"};
    Solution obj;

    obj.groupAnagrams(strs);

    return 0;
}